Fix language match on Linux (Unix?)

eatradish opened the pull request:
In Linux, it is possible for the `LANGUAGE` variable to exist but have a empty value,

Before: 

![图片](https://github.com/user-attachments/assets/1c9971b6-f612-421e-a915-96ad9971ed72)

After:

![图片](https://github.com/user-attachments/assets/5ddc708d-27e6-4cbd-9699-bfef84e0fb35)

eatradish commented:
@microsoft-github-policy-service agree

eatradish commented:
cc @Kijewski https://github.com/microsoft/edit/pull/104/commits/e35b6204add804b5c4b0cbb87379c2c8589c9e30 You're  right, I've reissued a version of the patch as per your statement

Kijewski commented:
@eatradish, oops, I just noticed that the territory must not be stripped: <https://github.com/microsoft/edit/blob/e8d40f6e7a95a6e19765ff19621cf0d39708f7b0/src/bin/edit/localization.rs#L934-L937>

Could you replace the stripping of everything after `_` with replacing `_` with `-`, please?

eatradish commented:
> @eatradish, oops, I just noticed that the territory must not be stripped:
> 
> https://github.com/microsoft/edit/blob/e8d40f6e7a95a6e19765ff19621cf0d39708f7b0/src/bin/edit/localization.rs#L934-L937
> 
> Could you replace the stripping of everything after `_` with replacing `_` with `-`, please?

I agree, but I think it's equally important to add the value you get from the '-' split to the result as well?

Kijewski commented:
Sorry, I did not express myself properly (by using the word "replace" twice):
- don't split at '-'
- instead make the underscore '_' a dash '-'

eatradish commented:
> @eatradish, oops, I just noticed that the territory must not be stripped:
> 
> https://github.com/microsoft/edit/blob/e8d40f6e7a95a6e19765ff19621cf0d39708f7b0/src/bin/edit/localization.rs#L934-L937
> 
> Could you replace the stripping of everything after `_` with replacing `_` with `-`, please?



> Sorry, I did not express myself properly (by using the word "replace" twice):
> 
>     * don't split at '-'
> 
>     * instead make the underscore '_' a dash '-'

I changed it to this, but for the final result I still have to split the '-' to successfully match  'zh':

```rust
pub fn preferred_languages(arena: &Arena) -> Vec<ArenaString<'_>, &Arena> {
    let mut locales = Vec::new_in(arena);

    for key in ["LANGUAGE", "LC_ALL", "LANG"] {
        if let Ok(langs) = std::env::var(key) {
            let langs = langs.replace('_', "-").to_ascii_lowercase();

            let langs = langs
                // split colon separated entries
                .split(':')
                // strip codeset ("utf8" in "C.utf8")
                .map(|s| s.split('.').next().unwrap_or(s))
                // filter out empty entries
                .filter(|s| !s.is_empty())
                // intern instrings in arena
                .map(|s| ArenaString::from_str(arena, s));
            locales.extend(langs);
            if !locales.is_empty() {
                break;
            }
        }
    }

    dbg!(&locales);

    locales
}

```
[src/sys/unix.rs:548:5] &locales = [
    "zh-cn",
]
```

```rust
"zh" => LangId::zh_hans, 
```

eatradish commented:
Maybe it should be changed to this？

```rust
pub fn preferred_languages(arena: &Arena) -> Vec<ArenaString<'_>, &Arena> {
    let mut locales = Vec::new_in(arena);

    for key in ["LANGUAGE", "LC_ALL", "LANG"] {
        if let Ok(langs) = std::env::var(key) {
            let langs_str = langs.replace('_', "-").to_ascii_lowercase();

            let langs = langs_str
                // split colon separated entries
                .split(':')
                // strip codeset ("utf8" in "C.utf8")
                .map(|s| s.split('.').next().unwrap_or(s))
                // filter out empty entries
                .filter(|s| !s.is_empty())
                // intern instrings in arena
                .map(|s| ArenaString::from_str(arena, s));

            locales.extend(langs);

            let Some((lang, _)) = langs_str.split_once('-') else {
                continue;
            };

            locales.push(ArenaString::from_str(arena, lang));

            if !locales.is_empty() {
                break;
            }
        }
    }

    dbg!(&locales);

    locales
}
```

lhecker commented on src/sys/unix.rs:
```diff
@@ -526,11 +526,22 @@ pub fn preferred_languages(arena: &Arena) -> Vec<ArenaString<'_>, &Arena> {
+            let langs = langs
+                // split colon separated entries
+                .split(':')
+                // strip territory ("EN" in "en_EN")
+                .map(|s| s.split('_').next().unwrap_or(s))
+                // strip codeset ("utf8" in "C.utf8")
+                .map(|s| s.split('.').next().unwrap_or(s))
```
This isn't the right place to do this. The reason matching doesn't work right now is because of #74. The issue shows where the fix needs to be made (= needs prefix match).

The only change needed in this function IMO is swapping the unconditional `break` with the `if` condition you added below.

lhecker replied:
I just came across #85 which actually almost fixes this bug (only the case-insensitivity is missing). As such, I think this PR only needs to fix the early-exit?

eatradish replied:
> I just came across #85 which actually almost fixes this bug (only the case-insensitivity is missing). As such, I think this PR only needs to fix the early-exit?

Done

lhecker replied:
Hmm that's odd. I'll try to test both of your PRs locally as soon as I can.

lhecker commented:
FYI we squash all our PRs (to keep our history clean). You can absolutely feel free to just keep pushing commits into this PR. 🙂

eatradish commented:
> FYI we squash all our PRs (to keep our history clean). You can absolutely feel free to just keep pushing commits into this PR. 🙂

hmm... Mostly it seems like all the problems I want fixed should be fixed in #85

lhecker reviewed (approved):
Thank you for fixing this! I'll try to merge in the other PR as soon as I can.

lhecker commented:
...I now realize why you replaced `_` with `-`. I apologize for that. 🥴
I'll fix this for #85.
